Google.Api.Gax.Grpc/ChannelPool.cs

```diff
@@ -76,14 +76,17 @@ private async Task<ChannelCredentials> CreateChannelCredentialsUncached()
         /// <returns>A task which will complete when all the (current) channels have been shut down.</returns>
         public Task ShutdownChannelsAsync()
         {
-            List<Channel> channelsToShutdown;
+            List<ChannelBase> channelsToShutdown;
             lock (_lock)
             {
                 channelsToShutdown = _channels.Values.ToList();
                 _channels.Clear();
             }
+            // FIXME: Reinstate this
+            return Task.CompletedTask;
+            /*
             var shutdownTasks = channelsToShutdown.Select(c => c.ShutdownAsync());
-            return Task.WhenAll(shutdownTasks);
```

Comment:
a bit more detail on shutting down ChannelBase:

- Channel (from Grpc.Core) has ShutdownAsync() method ( https://github.com/grpc/grpc/blob/528dbb13e400654f9c73d8e26e70be48f27052c4/src/csharp/Grpc.Core/Channel.cs#L224)
- GrpcChannel (from grpc-dotnet) has `Dispose()` method (which disposes the underlying HttpClient if configured so https://github.com/grpc/grpc-dotnet/blob/36e42811e7bf85c8aa765813c82fab3dcfeb8297/src/Grpc.Net.Client/GrpcChannel.cs#L270)
- ChannelBase does not expose either of those. I think we could add either ShutdownAsync() or Dispose() to ChannelBase to provide some way of signaling that the channel is no longer going to be used.

The managed GrpcChannel is a thin wrapper on top of HttpClient. I wasn't able to find much info on how important it is to Dispose() HttpClient when done with it.

I filed https://github.com/grpc/grpc/issues/21612 to track this.

Reply:
What if implement the same solution as for instantiation of the client - pass a delegate do shutdown the unwanted clients?

Reply:
That would definitely be possible. We could potentially isolate everything into one abstract class. It all feels somewhat weird, admittedly.
The bigger issue we need to address is what the final set of dependencies is. Will write more about that later.

Reply:
See the comment on ChannelFactory for deployment thoughts.

Reply:
I think that adding ShutdownAsync to ChannelBase is fine. When I added ChannelBase I considered adding some of Channel's methods to it (state, connect, shutdown) but since we didn't have a reason for them I left ChannelBase bare-bones.

What you can do:
* Move ShutdownAsync from Channel to ChannelBase
* Add virtual ShutdownAsyncCore to ChannelBase. By default it does nothing.
* Override ShutdownAsyncCore on Channel to run its current ShutdownAsync logic.
* Override ShutdownAsyncCore on GrpcChannel to call HttpClient.Dispose.